cpp/src/io/parquet/compact_protocol_reader.cpp

```diff
@@ -502,6 +502,24 @@ class parquet_field_optional : public parquet_field {
   }
 };
 
+/**
+ * @brief functor to wrap functors for optional fields
+ */
+template <typename T, typename FieldFunctor>
+class parquet_field_device_optional : public parquet_field {
```

Comment:
This is needed since we currently rely on relaxed constexpr so `std::optional` can be used on the device.

A question for the Parquet code owners: Should we update `parquet_field_optional` to use `cuda::std::optional`, or create a separate class dedicated to handling host-device optional fields like here?

Reply:
Pretty sure that the `std::optional` use on the device was unintentional.
I'm fine with using `cuda::std::optional`
CC @etseidl in case I'm misrepresenting the original design.

Reply:
looks like the use of `std::optional` was introduced by #17015. It was always intended to _not_ use `std::optional` because some thrift structs are used on device (like `LogicalType`). I'd probably prefer reverting to `cuda::std::optional` here.

Reply:
Well, according to the description of https://github.com/rapidsai/cudf/pull/17015, we can't use `cuda::std::optional` for stuff like `std::vector`s, e.g. `encoding_stats`. I'll take a closer look at the new changes.

Reply:
Well, we could also convert thrift to something more device friendly, rather than just relying on the thrift defined structs. I did a quick peek at the changes, and am now wondering if using `LogicalType` on device was a desire, but I left things using `ConvertedType` for now. Gotta reawaken all those neurons...my brain is full of Rust these days :P

Reply:
I've reverted this change and will create a separate PR to follow up.